app/app/src/main/kotlin/com/hedvig/app/feature/chat/data/ChatRepositoryNew.kt

```diff
@@ -0,0 +1,120 @@
+package com.hedvig.app.feature.chat.data
+
+import arrow.core.Either
+import arrow.core.raise.either
+import com.apollographql.apollo3.ApolloClient
+import com.hedvig.android.apollo.safeExecute
+import com.hedvig.android.apollo.toEither
+import com.hedvig.android.core.common.ErrorMessage
+import com.hedvig.android.logger.LogPriority
+import com.hedvig.android.logger.logcat
+import kotlinx.datetime.Instant
+import octopus.ChatMessagesQuery
+import octopus.ChatSendFileMutation
+import octopus.ChatSendMessageMutation
+import octopus.fragment.MessageFragment
+import octopus.type.ChatMessageFileInput
+import octopus.type.ChatMessageSender
+import octopus.type.ChatMessageTextInput
+
+interface ChatRepositoryNew {
+  suspend fun fetchChatMessages(until: Instant? = null): Either<ErrorMessage, ChatMessagesResult>
+  suspend fun sendFile(uploadUrl: String): Either<ErrorMessage, ChatMessageResult>
+  suspend fun sendMessage(text: String): Either<ErrorMessage, ChatMessageResult>
+}
+
+class ChatRepositoryNewImpl(
+  private val apolloClientOctopus: ApolloClient,
+) {
+  suspend fun fetchChatMessages(until: Instant? = null) = either {
+    val result = apolloClientOctopus.query(ChatMessagesQuery(until))
+      .safeExecute()
+      .toEither(::ErrorMessage)
+      .bind()
+
+    ChatMessagesResult(
+      messages = result.chat.messages.mapNotNull { it.toMessage() },
+      nextUntil = result.chat.nextUntil,
+      hasNext = result.chat.hasNext,
+    )
+  }
+
+  suspend fun sendFile(uploadUrl: String) = either {
+    val result = apolloClientOctopus.mutation(ChatSendFileMutation(ChatMessageFileInput(uploadUrl)))
+      .safeExecute()
+      .toEither(::ErrorMessage)
+      .bind()
+
+    val error = result.chatSendFile.error
+    val message = result.chatSendFile.message
+    val status = result.chatSendFile.status
+
+    if (error != null) {
+      raise(ErrorMessage(error.message))
+    } else if (message == null) {
+      raise(ErrorMessage("No data"))
+    } else {
+      ChatMessageResult(
+        message = message.toMessage(),
+        status = status?.message ?: "",
+      )
+    }
+  }
+
+  suspend fun sendMessage(text: String) = either {
+    val result = apolloClientOctopus.mutation(ChatSendMessageMutation(ChatMessageTextInput(text)))
+      .safeExecute()
+      .toEither(::ErrorMessage)
+      .bind()
+
+    val error = result.chatSendText.error
+    val message = result.chatSendText.message
+    val status = result.chatSendText.status
+
+    if (error != null) {
+      raise(ErrorMessage(error.message))
+    } else if (message == null) {
+      raise(ErrorMessage("No data"))
+    } else {
+      ChatMessageResult(
+        message = message.toMessage(),
+        status = status?.message ?: "",
+      )
+    }
```

Comment:
Ensure is pretty convenient in the context of an `either` as it just lets you take care of some scenarios one by one, without having to nest anything in if checks.
It'd look something like this here
```suggestion
    ensure(error == null) {
      ErrorMessage(error.message)
    }
    ensureNotNull(message) {
      ErrorMessage("No data")
    }
    ChatMessageResult(
      message = message.toMessage(),
      status = status?.message ?: "",
    )
```
Where you ensure that error is null, otherwise it raises the error and returns early.
Then it ensures that message is not null, early-returning again with the correct ErrorMessage otherwise.
And in the end, you're just left with the good path, and `error` and `message` are smart cast at this point to be null and non-null accordingly.

This way you avoid having to call `raise` completely, and the code reads a tiny bit easier without the chained if statement I think.

But this is a stylistic change, just sharing my opinion.